### PR TITLE
Upload makefile2.0

### PR DESCRIPTION
**Pull request by MikeDrewitt**
This branch adds the ability to upload an autograder makefile and/ or tar from the front end autograder settings. The two upload buttons are added under the autograder editor form. The uploads will be moved into the corresponding course/ assignment directory with the correct name. The files can then be used can then used to grade assignments. 

The reason we added this feature was to allow anyone managing an assignment to edit, or upload the required files for said assignment without needing access to the grading server.

### CODE REVIEW

**Comment by nayak16** [comment]
Can you please post screenshots just for documentation?

**Comment by nayak16** [comment]
I'm getting errors with rake v12.0.0 when trying to test this branch out. Have you updated rake or are using a higher version?

File: app/views/autograders/_form.html.erb [hunk 1]
```diff
@@ -1,4 +1,28 @@
-<%= form_for [@course, @assessment, @autograder], builder: FormBuilderWithDateTimeInput do |f| %>
+<% content_for :javascripts do %>
+	<script>
+		$(function() {
+		  $(document).on('change', ':file', function() {
+		    var input = $(this),
+		    	numFiles = input.get(0).files ? input.get(0).files.length : 1,
+		      label = input.val().replace(/\\/g, '/').replace(/.*\//, '');
+		    	input.trigger('fileselect', [numFiles, label]);
```
Comment (nayak16) [inline app/views/autograders/_form.html.erb]: Can you fix the tab levels here?

File: app/views/autograders/_form.html.erb [hunk 2]
```diff
@@ -1,4 +1,28 @@
...
+		  });
+
+		  $(document).ready( function() {
+		    $('#autograder_makefile:file').on('fileselect', function(event, numFiles, label) {
```
Comment (nayak16) [inline app/views/autograders/_form.html.erb]: Why not just use the default [Materialize File Input](http://materializecss.com/forms.html)? It does this for you?

File: app/controllers/autograders_controller.rb [hunk 3]
```diff
@@ -25,15 +25,32 @@ def edit
...
     redirect_to([:edit, @course, @assessment]) && return
   end
 
+  action_auth_level :upload, :instructor
+  def upload
+    uploaded_makefile = params[:autograder][:makefile]
+	  uploaded_tar = params[:autograder][:tar]
+	  if not uploaded_makefile.nil?
+		  File.open(Rails.root.join('courses', @course.name, @assessment.name, 'autograde-Makefile'), 'wb') do |file|
+		  file.write(uploaded_makefile.read) unless uploaded_makefile.nil?
```
Comment (nayak16) [inline app/controllers/autograders_controller.rb]: indent

File: app/views/autograders/_form.html.erb [hunk 4]
```diff
@@ -10,6 +34,26 @@
   <%= f.check_box :release_score, display_name: "Release Scores?",
     help_text: "Check to release autograded scores to students immediately after autograding (strongly recommended)." %>
 
+	<br />
+	<label class="btn primary">
+  	Autograder Makefile
+	  <span class="hide-all">
+  	  <%= f.file_field :makefile, action: :upload %>
```
Comment (nayak16) [inline app/views/autograders/_form.html.erb]: Again, why not just use default file field, it includes the button.

File: Gemfile [hunk 5]
```diff
@@ -1,6 +1,6 @@
 source 'https://rubygems.org'
 
-gem 'rails', '>=4.2.0'
+gem 'rails', '=4.2.1'
```
Comment (esse) [inline Gemfile]: why freezing rails version?

File: Gemfile [hunk 6]
```diff
@@ -42,14 +42,14 @@ gem 'rake', '>=10.3.2'
 gem 'populator', '>=1.0.0'
 
 # To communicate with MySQL database
-gem 'mysql2'
+gem 'mysql2', '=0.3.18'
```
Comment (esse) [inline Gemfile]: why freeze?

File: app/controllers/autograders_controller.rb [hunk 7]
```diff
@@ -25,15 +25,32 @@ def edit
   action_auth_level :update, :instructor
   def update
     flash[:info] = "Saved!" if @autograder.update(autograder_params)
+		upload
```
Comment (esse) [inline app/controllers/autograders_controller.rb]: bad indentation

File: app/controllers/autograders_controller.rb [hunk 8]
```diff
@@ -25,15 +25,32 @@ def edit
...
   def destroy
     flash[:info] = "Destroyed!" if @autograder.destroy
     redirect_to([:edit, @course, @assessment]) && return
   end
 
+  action_auth_level :upload, :instructor
+  def upload
+    uploaded_makefile = params[:autograder][:makefile]
+	  uploaded_tar = params[:autograder][:tar]
+	  if not uploaded_makefile.nil?
```
Comment (esse) [inline app/controllers/autograders_controller.rb]: don't use not. Instead use `!`.

and even better, use `unless` instead (`unless condition` is the same as `if !condition`)

File: app/controllers/autograders_controller.rb [hunk 9]
```diff
@@ -25,15 +25,32 @@ def edit
...
+		  File.open(Rails.root.join('courses', @course.name, @assessment.name, 'autograde-Makefile'), 'wb') do |file|
+		  file.write(uploaded_makefile.read) unless uploaded_makefile.nil?
+		end	
+	  end
+	  if not uploaded_tar.nil?
```
Comment (esse) [inline app/controllers/autograders_controller.rb]: use unless instead (unless condition is the same as if !condition)

File: app/controllers/autograders_controller.rb [hunk 10]
```diff
@@ -25,15 +25,32 @@ def edit
...
+    uploaded_makefile = params[:autograder][:makefile]
+	  uploaded_tar = params[:autograder][:tar]
+	  if not uploaded_makefile.nil?
+		  File.open(Rails.root.join('courses', @course.name, @assessment.name, 'autograde-Makefile'), 'wb') do |file|
+		  file.write(uploaded_makefile.read) unless uploaded_makefile.nil?
+		end	
+	  end
+	  if not uploaded_tar.nil?	
+		  File.open(Rails.root.join('courses', @course.name, @assessment.name, 'autograde.tar'), 'wb') do |file|
+		  file.write(uploaded_tar.read) unless uploaded_tar.nil?
```
Comment (esse) [inline app/controllers/autograders_controller.rb]: indent

**Comment by nayak16** [comment]
Closing. This functionality is part of #776

**Pull request closed without merging**